pkg/download/list.go

```diff
@@ -1,35 +1,150 @@
 package download
 
 import (
+	"context"
+	"errors"
 	"net/http"
 	"strings"
 
+	"github.com/sirupsen/logrus"
+
+	"github.com/gomods/athens/pkg/log"
+
 	"github.com/bketelsen/buffet"
 	"github.com/gobuffalo/buffalo"
 	"github.com/gobuffalo/buffalo/render"
+	"github.com/gomods/athens/pkg/config/env"
 	"github.com/gomods/athens/pkg/paths"
 	"github.com/gomods/athens/pkg/storage"
+	"github.com/google/go-github/github"
+	"github.com/marwan-at-work/vgop/semver"
 	errs "github.com/pkg/errors"
 )
 
 // PathList URL.
 const PathList = "/{module:.+}/@v/list"
 
 // ListHandler implements GET baseURL/module/@v/list
-func ListHandler(lister storage.Lister, eng *render.Engine) func(c buffalo.Context) error {
+func ListHandler(
+	lister storage.Lister,
+	eng *render.Engine,
+	isProxy bool,
+	lggr *log.Logger,
+) func(c buffalo.Context) error {
 	return func(c buffalo.Context) error {
 		sp := buffet.SpanFromContext(c)
 		sp.SetOperationName("listHandler")
 		mod, err := paths.GetModule(c)
 		if err != nil {
 			return err
 		}
+
 		versions, err := lister.List(mod)
-		if storage.IsNotFoundError(err) {
-			return c.Render(http.StatusNotFound, eng.JSON(err.Error()))
+		if storage.IsNotFoundError(err) || len(versions) == 0 {
+			if isProxy {
+				// Go uses the http.DefaultClient which implicitly handles up to 10 reqdirects.
+				// therefore, it's safe to redirect to whatever server/endpoint we like as long
+				// as that server returns an answer that Go expects.
+				return c.Redirect(http.StatusMovedPermanently, env.GetOlympusEndpoint())
```

Comment:
@marwan-at-work do you happen to know if the redirect is "officially supported" in `go get`? I'd like to make sure it's a somewhat properly specified feature of `go get` and we're not relying on an implementation detail in the CLI

Reply:
as long as `go get` uses the standard library `net/http` DefaultClient (which it does) redirects are officially supported.